File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
import sys

strN,strK = sys.stdin.readline().split()
N =int(strN)
K= int(strK)

#DP
Nlist = [[0,0]]

for i in range(N):
    Nlist.append(list(map(int,sys.stdin.readline().split())))


dp = [[0] * (K+1) for _ in range(N+1)]

for i in range(1,N+1) :

    for j in range(1,K+1) :

        weight= Nlist[i][0]
        value= Nlist[i][1]

        if j<weight:
            dp[i][j] = dp[i-1][j]
        else:
            dp[i][j]= max(dp[i-1][j],dp[i-1][j-weight]+value)

print(dp[N][K])

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.java
    import java.util.*;
    import java.lang.*;
    import java.io.*;
    
    // The main method must be in a class named "Main".
    class Main {
        public static void main(String[] args) throws IOException{
            BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
            BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    
            String [] oneline = br.readLine().split(" ");
            int N = Integer.parseInt(oneline[0]);
            int K = Integer.parseInt(oneline[1]);
            int [] valvarr = new int [N];
            int [] weightarr = new int[N];
    
            for (int i = 0 ; i<N; i++)
                {
                        String [] twoline = br.readLine().split(" ");
                    int weight = Integer.parseInt(twoline[0]); 
                    int val = Integer.parseInt(twoline[1]); 
                    valvarr[i] = val;
                    weightarr[i] = weight;
                }

            int [][] dp = new int [N][K+1];

            for (int j = 0 ; j<=K; j++)
                {
                    if (j >= weightarr[0])
                    {
                        dp[0][j] = valvarr[0];
                    }
                }

            for (int i = 1; i<N; i++)
                {
                    for (int j = 0 ; j<=K; j++)
                        {
                            dp[i][j] = dp[i-1][j];

                            if (weightarr[i] <= j)
                            {
                                dp[i][j]= Math.max(dp[i][j],dp[i-1][j-weightarr[i]]+valvarr[i]);
                            }
                        }
                }

            bw.write(String.valueOf(dp[N-1][K]));
            bw.close();
        }
    }